src/backend/access/transam/xlog.c

```diff
@@ -8811,29 +8865,106 @@ GetInsertRecPtr(void)
 }
 
 /*
- * GetLastWrittenPageLSN -- Returns maximal LSN of written page
+ * GetLastWrittenLSN -- Returns maximal LSN of written page.
+ * It returns either cached last written LSN of particular relation,
+ * either global maximum of last written LSNs among all relations.
  */
 XLogRecPtr
-GetLastWrittenPageLSN(void)
+GetLastWrittenLSN(Oid rnode, BlockNumber blkno)
 {
 	XLogRecPtr lsn;
-	SpinLockAcquire(&XLogCtl->info_lck);
-	lsn = XLogCtl->lastWrittenPageLSN;
-	SpinLockRelease(&XLogCtl->info_lck);
+	LastWrittenLsnCacheEntry* entry;
+
+	LWLockAcquire(LastWrittenLsnLock, LW_SHARED);
+
+	/* Maximal last written LSN among all non-cached pages */
+	lsn = XLogCtl->maxLastWrittenLsn;
+
+	if (rnode != InvalidOid)
+	{
+		LastWrittenLsnCacheKey key;
+		key.relid = rnode;
+		key.basket = blkno / LAST_WRITTEN_LSN_CACHE_BASKET;
+		entry = hash_search(lastWrittenLsnCache, &key, HASH_FIND, NULL);
+		if (entry != NULL)
+			lsn = entry->lsn;
+	}
+	else
+	{
+		HASH_SEQ_STATUS seq;
+		/* Find maximum of all cached LSNs */
+		hash_seq_init(&seq, lastWrittenLsnCache);
+		while ((entry = (LastWrittenLsnCacheEntry *) hash_seq_search(&seq)) != NULL)
+		{
+			if (entry->lsn > lsn)
+				lsn = entry->lsn;
+		}
+	}
+	LWLockRelease(LastWrittenLsnLock);
 
 	return lsn;
 }
 
 /*
- * SetLastWrittenPageLSN -- Set maximal LSN of written page
+ * SetLastWrittenLSN -- Set maximal LSN of written page.
+ * We maintain cache of last written LSNs with limited size and LRU replacement
+ * policy. To reduce cache size we store max LSN not for each page, but for
+ * backet (1024 blocks). This cache allows to use old LSN when
+ * requesting pages of unchanged or appended relations.
  */
 void
-SetLastWrittenPageLSN(XLogRecPtr lsn)
+SetLastWrittenLSN(XLogRecPtr lsn, Oid rnode, BlockNumber from, BlockNumber till)
```

Comment:
rnode isn't enough to uniquely identify a relation. If you have two tables with same relfilenode, what happens?

Also, what about FSM and VM forks?

Reply:
I have explained it in comment to LastWrittenLsnCacheKey struct: I just want to make key shorter.
Probabiity of conflict (i.e. that two different relations from different databases has the same oid and are together updated) is relatively small. Concern FSM/VM forks, it is even better that them are placed in the same backet, because them are accessed together.

In  any case: this mapping is not required to be unique: we store max LSN. This is why it is possible to store single LSN for range of pages (and forks/relations).